### PR TITLE
jetese analysis framework

### PR DESCRIPTION
**Pull request by joachimckh**

### CODE REVIEW

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (nzardosh) [inline PWGJE/Tasks/jetSpectraESE.cxx:313]: you should make sure the deltaPhi is conatrained

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (nzardosh) [inline PWGJE/Tasks/jetSpectraESE.cxx:206]: Is this the only piece of information from the original track that you need?

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (nzardosh) [inline PWGJE/Tasks/jetSpectraESE.cxx:155]: What information from the original collision is missing in JCollisions to make this work?

**Review by nzardosh — COMMENTED** [review]
Hi joachim,

Thanks for your PR. I see your task takes a bunch of different process functions from the skeleton task such as angularity or hadron jet recoils. These are two quite different analyses. What do you plan to measure with your task?

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (joachimckh) [reply PWGJE/Tasks/jetSpectraESE.cxx:155]: What do you mean exactly? I have joined JCollisions with my own tables in the top:
using JColwESE = soa::Join<aod::JCollisions, aod::CentFT0Cs, aod::qVecFT0Cs, aod::qPercentileFT0Cs>;
this shoudl work, no?

Comment (nzardosh) [reply PWGJE/Tasks/jetSpectraESE.cxx:155]: Thanks for the quick answer. Yes for now it would be great if you just keep the process functions you need and you can always add more later :)
All the best,
Nima

Comment (nzardosh) [reply PWGJE/Tasks/jetSpectraESE.cxx:155]: Ah I see, i think i misread the collisions. Ok that is fine then

Comment (joachimckh) [reply PWGJE/Tasks/jetSpectraESE.cxx:107]: Thanks! I have removed all other process functions for now. :)
Cheers,
Joachim

**Comment by victor-gonzalez** [comment]
@ddobrigk As the Q vector part seems to expand the Event Plane framework I would suggest to place it at the same level
Perhaps existing functionality from that framework, which is already tested, can be even reused instead

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (nzardosh) [inline PWGJE/Tasks/jetSpectraESE.cxx:42]: just an additonal comment on this. In general since JCollisions is a (self contained) derived data format, its not foreseen to have it always joinable tables that are at their core joinable with the aod::Collisions table. For now if you prefer to keep this this way it is fine but it can happen that some development we do makes JCollisions and Collisions not joinable anymore. The safer option would be to have in your process function process(soa::Join<JetCollisions, aod::JCollisionPIs>::iterator const& collision, soa::Join<aod::Collisions, aod::CentFT0Cs, aod::qVecFT0Cs, aod::qPercentileFT0Cs> const& ) and then inside this function do auto originalCollision = collision.collision_as<soa::Join<aod::Collisions, aod::CentFT0Cs, aod::qVecFT0Cs, aod::qPercentileFT0Cs>>() and then get the qvector etc from originalCollision

Comment (joachimckh) [reply PWGJE/Tasks/jetSpectraESE.cxx:42]: Thanks a lot! I will change it in a future commit (I don't think it's important to change now to make this pr).

**Review by victor-gonzalez — CHANGES_REQUESTED** [review]
The Q vector functionality has to be placed with the Event plane and Q vector functionality under `Common`

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (nzardosh) [reply PWGJE/Tasks/jetSpectraESE.cxx:42]: ok but just to let you know that your task might stop working at some point unexpectedly due to this

Comment (nzardosh) [reply PWGJE/Tasks/jetSpectraESE.cxx:42]: since you are making other changes anyways you could consider incorporating it

Comment (joachimckh) [reply PWGJE/Tasks/jetSpectraESE.cxx:42]: I have changed it now, since Victor requested me to relocate some of the code. So it will make the next commit. Thanks for the suggestions. :)

**Comment by joachimckh** [comment]
Hi Victor, I have relocated some of the code (failing some trivial tests I'm changing). However, my weight class (FFitWeights) now have to be linked with all analyses (compared to just a few before), which is a bit tedious whenever I am making changes, is there any way for me to avoid this? If I don't link it in AnalysisCore, the compilation fails.

Cheers,
Joachim

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (vkucera) [inline PWGJE/Tasks/jetSpectraESE.cxx:147]: Please don't use `TaskName` if not needed by templatation. If you just name your struct `JetSpectraEse`, you will get the same result.

**Comment by vkucera** [comment]
@joachimckh Please note that you don't need to format your code manually. See the linked automatic PRs.

**Comment by joachimckh** [comment]
Hi @vkucera, I used clang-format on the files. I had some issues with the order of the header files, resulting in the larger number of commits. Is there a better way to do it?

Cheers,
Joachim

**Comment by vkucera** [comment]
> Hi @vkucera, I used clang-format on the files. I had some issues with the order of the header files, resulting in the larger number of commits. Is there a better way to do it?
> 
> Cheers, Joachim

You can completely avoid running `clang-format` by just merging the automatic formatting PR. See details at https://aliceo2group.github.io/analysis-framework/docs/gettingstarted/contributingtocode.html#pull-requests

Order of header includes is a separate issue which is checked by MegaLinter.

**Review by nzardosh — APPROVED** [review]

**Comment by joachimckh** [comment]
I see, thanks a lot! I will try that next time @vkucera

File: PWGCF/Flow/CMakeLists.txt
Comment (victor-gonzalez) [inline PWGCF/Flow/CMakeLists.txt:15]: This file should not be affected

File: PWGCF/Flow/Core/CMakeLists.txt
Comment (victor-gonzalez) [inline PWGCF/Flow/Core/CMakeLists.txt:11]: This file should not be affected

File: PWGCF/Flow/DataModel/CMakeLists.txt
Comment (victor-gonzalez) [inline PWGCF/Flow/DataModel/CMakeLists.txt:10]: This file should not be included

File: PWGCF/Flow/TableProducer/CMakeLists.txt
Comment (victor-gonzalez) [inline PWGCF/Flow/TableProducer/CMakeLists.txt:11]: This file should not be affected

**Review by victor-gonzalez — CHANGES_REQUESTED** [review]
Please, remove the changes that affect `PWGCF` files. They would create the false impression that this PR involves them, which is not the case

**Comment by victor-gonzalez** [comment]
> Hi Victor, I have relocated some of the code (failing some trivial tests I'm changing). However, my weight class (FFitWeights) now have to be linked with all analyses (compared to just a few before), which is a bit tedious whenever I am making changes, is there any way for me to avoid this? If I don't link it in AnalysisCore, the compilation fails.
> 
> Cheers, Joachim

In my knowledge, the only thing which will decrease the changes impact is expend a bit more time in defining the header files in a way that they will not change so often. Only re-linking would be needed in that case for other changes that don't involve the header files

File: PWGCF/Flow/CMakeLists.txt
Comment (joachimckh) [reply PWGCF/Flow/CMakeLists.txt]: Hi Victor, I have restored all the PWGCF files to origin/master

**Comment by joachimckh** [comment]
> > Hi Victor, I have relocated some of the code (failing some trivial tests I'm changing). However, my weight class (FFitWeights) now have to be linked with all analyses (compared to just a few before), which is a bit tedious whenever I am making changes, is there any way for me to avoid this? If I don't link it in AnalysisCore, the compilation fails.
> > Cheers, Joachim
> 
> In my knowledge, the only thing which will decrease the changes impact is expend a bit more time in defining the header files in a way that they will not change so often. Only re-linking would be needed in that case for other changes that don't involve the header files

Okay, thanks for the reply. The weights class might need further modifications in the future, but it should be mostly fully fletched out currently, so I suppose it's not that big of an issue.

**Review by victor-gonzalez — APPROVED** [review]
Thanks!
I approve what is on my side!

File: Common/TableProducer/FlowESETable.cxx
Comment (vkucera) [inline Common/TableProducer/FlowESETable.cxx:1]: The file name suggests that this file defines a table which is not the case. Please rename it to something that clearly indicates that this code produces the table.
Also please name the `struct` inside accordingly so that there is a clear correspondence between the device name (appearing in the log) and the name of the file that implements it.

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (vkucera) [inline PWGJE/Tasks/jetSpectraESE.cxx:1]: Similar comment on naming. Please rename the file to `jetSpectraEseTask.cxx`

Comment (joachimckh) [reply PWGJE/Tasks/jetSpectraESE.cxx:1]: @vkucera I changed the names now, I hope it makes better sense

Comment (vkucera) [reply PWGJE/Tasks/jetSpectraESE.cxx:1]: @joachimckh Thanks, that looks much clearer.

File: Common/TableProducer/ProduceEseTable.cxx
Comment (vkucera) [inline Common/TableProducer/ProduceEseTable.cxx:1]: Last minor fix. For consistency with the other files, please change `ProduceEseTable.cxx` to `produceEseTable.cxx`.

File: Common/TableProducer/CMakeLists.txt
Comment (vkucera) [reply Common/TableProducer/CMakeLists.txt:117]: Sorry, obsolete comment.

File: PWGJE/Tasks/jetSpectraESE.cxx
Comment (vkucera) [reply PWGJE/Tasks/jetSpectraESE.cxx]: Thanks

File: Common/TableProducer/CMakeLists.txt
Comment (vkucera) [inline Common/TableProducer/CMakeLists.txt:117]:
```suggestion
                    SOURCES eseTableProducer.cxx
```

Comment (vkucera) [inline Common/TableProducer/CMakeLists.txt:117]:
```suggestion
o2physics_add_dpl_workflow(ese-table-producer
                    SOURCES eseTableProducer.cxx
```

Comment (joachimckh) [reply Common/TableProducer/CMakeLists.txt]: Should be changed now along with the other one. I had a bit of git issues hence the way I implemented suggestions..

File: PWGJE/Tasks/CMakeLists.txt
Comment (vkucera) [inline PWGJE/Tasks/CMakeLists.txt:172]:
```suggestion
    o2physics_add_dpl_workflow(jet-spectra-ese-task
                        SOURCES jetSpectraEseTask.cxx
```

File: Common/TableProducer/CMakeLists.txt
Comment (vkucera) [reply Common/TableProducer/CMakeLists.txt]: Thanks

File: Common/DataModel/EseTable.h
Comment (vkucera) [inline Common/DataModel/EseTable.h:27]:
```suggestion
namespace q_vector
```
Please respect naming conventions.

File: Common/DataModel/EseTable.h
Comment (vkucera) [inline Common/DataModel/EseTable.h:41]:
```suggestion
DECLARE_SOA_TABLE(QVecFV0As, "AOD", "QVECFV0A", q_vector::QFV0ARe, q_vector::QFV0AIm);
DECLARE_SOA_TABLE(QVecFT0Cs, "AOD", "QVECFT0C", q_vector::QFT0CRe, q_vector::QFT0CIm);
DECLARE_SOA_TABLE(QPercentileFT0Cs, "AOD", "QPERCENTILEFT0C", q_vector::QPERCFT0C);
using QVecFV0A = QVecFV0As::iterator;
using QVecFT0C = QVecFT0Cs::iterator;
using QPercentileFT0C = QPercentileFT0Cs::iterator;
```
Please respect naming conventions.

File: Common/DataModel/EseTable.h
Comment (vkucera) [inline Common/DataModel/EseTable.h:35]:
```suggestion
} // namespace q_vector
```

**Review by nzardosh — APPROVED** [review]

File: Common/DataModel/EseTable.h
Comment (joachimckh) [reply Common/DataModel/EseTable.h:35]: Hi Vit, using the PR to commit suggestions is very smart. But my lack of git knowledge is giving me some quite big git issues now where I cannot push to my repository, since there's differences in files that are now modified several places. Would it be better to close this PR and make a new one?

Comment (joachimckh) [reply Common/DataModel/EseTable.h]: I force pushed changes now, but I should have added your suggestions..

Comment (vkucera) [reply Common/DataModel/EseTable.h]: Hi @joachimckh , please try to avoid overwriting the PR history with force-pushing. Details are explained in the guidelines for contributors: https://aliceo2group.github.io/analysis-framework/docs/gettingstarted/contributingtocode.html#pull-request-review
If commits are added to your branch on GitHub, you should sync your local branch with it using `git pull --rebase origin <branch>`, where in this case your branch is `master`. This will take the commit history from your remote branch and put your additional local commits on top of them so that you can cleanly push them to your remote branch. Again, more details are on the documentation pages I linked.

Comment (joachimckh) [reply Common/DataModel/EseTable.h]: @vkucera  Alright, thanks for all the information. I will avoid it in the future. 

The changes I put, is the renaming of the variable that you requested, as well as a small change in FFitWeights.

**Comment by joachimckh** [comment]
Hi @vkucera, can you help me, what more I need to do to get +1 for this?

Thanks a lot!

Cheers,
Joachim

**Comment by vkucera** [comment]
> Hi @vkucera, can you help me, what more I need to do to get +1 for this?
> 
> Thanks a lot!
> 
> Cheers, Joachim

Hi @joachimckh , you first need to resolve the conflicts and then you need approval from the code owners of the directories you touched.

**Comment by joachimckh** [comment]
Okay thanks a lot. However, there should not be any more conflicts, no?
I will be more specific and adhere to the guidelines in the future.

Cheers,
Joachin

**Comment by vkucera** [comment]
> Okay thanks a lot. However, there should not be any more conflicts, no? I will be more specific and adhere to the guidelines in the future.
> 
> Cheers, Joachin

Please see the list of conflicting files at the bottom of this page. See the linked instructions on how you can update your branch without overwriting it using `git merge`.

**Comment by joachimckh** [comment]
@vkucera I'm sorry, but I'm getting quite the issues now trying to merge things. I tried to branch out with development, to make sure nothing new is deleted, so locally I created another branch, and then switched branch to master. But I'm getting somewhat a git headache now, trying to follow the instructions.

So I did git pull --rebase origin master, and it shows me the conflicts for the CMakeLists files, which I have then tried to adjust. But now all my code is restored to old names (which I had to change according to this PR). I dont know if the code in there is updated or not, but I assume it's not. I tried to do the git merge upstream/master according to the documentation, but it just returns
git merge upstream/master not something we can merge

**Comment by victor-gonzalez** [comment]
@joachimckh I would recommend you to attend the DPG/AOT meeting this week (https://indico.cern.ch/e/1455087) 
They will address considerations around the current Q-vectors implementation

**Comment by victor-gonzalez** [comment]
@joachimckh You can even consider to give a presentation about your Q-vector functionality
I guess the owners of the centralized Q-vector functionality will be attending in case you need to ask for their approval

**Comment by joachimckh** [comment]
@victor-gonzalez thanks, I will put it in my calendar. I think this PR should be ready for approval now, but I'm not sure who?
@vkucera  helped me solve some of the git issues I was facing

**Comment by victor-gonzalez** [comment]
I guess the best way to get the approval is presenting the functionality (the Q-vactor one, and perhaps why it is needed) at that meeting and asking there for approval
Of course the PWGJE functionality can be approved by any of the people with rights on that PWG. You can ask to your PAG or PWG for that

**Comment by joachimckh** [comment]
@victor-gonzalez Okay, thank you! I will reach out there and ask to do that :)
Cheers,
Joachim

**Comment by ddobrigk** [comment]
Dear all: this PR has been hanging here for too long. I think @joachimckh already [presented at the DPG AOT](https://indico.cern.ch/event/1458453/) a while ago, and it seems he fulfilled everything that was asked of him, and from a look at the code, I believe this is OK to merge. Is there anything else? 

I will wait until the end of the day and will then grant `Common` approval. I ask that the JE conveners approve the JE part as well (@nzardosh and Laura Havener). Thanks!

**Review by nzardosh — APPROVED** [review]

**Review by ddobrigk — APPROVED** [review]

**Comment by ddobrigk** [comment]
@joachimckh I fear the compilation does not work... Can you please fix it? Thank you!!

**Comment by joachimckh** [comment]
> @joachimckh I fear the compilation does not work... Can you please fix it? Thank you!!

Hi David,

Thanks for the approval. It compiled last time I checked, but it has been hanging a while, so I will check what is changed when I get a chance. I will send a message when it will be ready.

Cheers, Joachim

**Comment by joachimckh** [comment]
Hi @ddobrigk and @nzardosh,
Please approve it again, it should compile again now!
Thanks a lot,
Joachim

File: PWGJE/Tasks/jetSpectraEseTask.cxx
Comment (vkucera) [inline PWGJE/Tasks/jetSpectraEseTask.cxx:156]: This code is completely unnecessary. Please consider replacing it with `RecoDecay::constrainAngle(mphi - vphi, -o2::constants::math::PI)`

File: Common/Core/FFitWeights.h
Comment (vkucera) [inline Common/Core/FFitWeights.h:15]: Please add documentation.

Comment (joachimckh) [reply Common/Core/FFitWeights.h]: Hi @vkucera,
I've added a short description in the header file for what it currently holds and what it will hold in the future. the source file refers to the header file now

File: PWGJE/Tasks/jetSpectraEseTask.cxx
Comment (joachimckh) [reply PWGJE/Tasks/jetSpectraEseTask.cxx]: @vkucera thanks for the pointer. I have changed the method here to use the RecoDecay function

Comment (vkucera) [reply PWGJE/Tasks/jetSpectraEseTask.cxx]: Thanks

File: Common/Core/FFitWeights.h
Comment (vkucera) [reply Common/Core/FFitWeights.h]: Thanks. Can you please format it using the Doxygen fields `\file`, `\author`, `\brief`, as described at https://rawgit.com/AliceO2Group/CodingGuidelines/master/comments_guidelines.html?showone=Files#Files ?

Comment (joachimckh) [reply Common/Core/FFitWeights.h]: @vkucera This should be done now :)

Comment (vkucera) [reply Common/Core/FFitWeights.h]: Thanks

**Review by ddobrigk — APPROVED** [review]

**Review by nzardosh — APPROVED** [review]

**Pull request merged**
